### ISSUE TITLE
Clarify local files IO

### ISSUE THREAD

**Issue by saket** [korlibs/korio#108]
I've been looking for a multiplatform library for files IO and I'm just realizing that `korio` is already capable of it. The term "virtual file system" was throwing me off. Is this something you'd be interested in clarifying by adding examples to the readme?

I could file issues to track completion of `Vfs#delete`, `Vfs#mkdir`, but the library looks in great shape overall. I also noticed that `korio` could auto-detect the current path instead of offloading it to the user (`File(".").absolutePath`).

**Comment by soywiz** [comment]
A virtual file system is an abstraction similar to File.
But that File could have different implementations: a memory-only filesystem, .ISO file, .ZIP file, S3, HTTP, SSH, FTP, disk, an encrypted file, etc.
This allows several scenarios: like doing testing by passing a memory filesystem constructed at runtime, or reusing the same code for dealing with a copy tree from disk, a zip file or whatever. Let's say it is more flexible that using a plain local filesystem.

**Comment by saket** [comment]
Right, I understand that now

**Comment by saket** [comment]
Slightly tangential question, that is again related to docs, what's the recommended way of reading files that does not load the entire content in memory? I'm using `VfsFile#readString()` but not sure if that's the best way.

**Comment by soywiz** [comment]
You can call the open method to get an AsyncStream then you can read chunks of a specific size and do stuff incrementally

**Comment by saket** [comment]
Gotcha. Could you share some codes? There are so many functions to read data from streams and I'm a bit overwhelmed.

**Comment by soywiz** [comment]
The other issue about MD5 includes a sample opening files and handling chunks for reading.

In the end you have `write` and `read` methods, `position` and `length`. The variants like `readU16BE`, are for reading primitive types on binary files.

If you have a specific thing you want to do, ley me know to point which methods to use

**Comment by saket** [comment]
Right, but just to confirm: does korio have any higher level abstractions around the low-level APIs you used in your MD5 changes? Something along the lines of [okio](https://github.com/square/okio):

```kotlin
file.source().buffer().use {
  it.readUtf8()
}
```

**Comment by soywiz** [comment]
Never used okio, so I'm not fully sure if I understand. But let me explain the design decisions of korio, that maybe help you to figure out yourself:

Consider the two approaches for handling streams asynchronously:
* read this amount of bytes, then execute the following code
* notify me when there is data and execute the following callback

The last approach have an issue with pressure. If you get more data than what you can consume, for example when copying to other place or pipeing, so you have to manually pause the stream when required. This was how node.js worked with streams at least in its first versions.

Korio is based on the first approach and is fully integrated with Kotlin's suspending functions. So you have to explicitly read stuff. That suspends, and then you do stuff. So if you want to consume the whole stream, that usually it is a while loop where you read.

We could add a function like `consume` or the like (not on the library right now IIRC), similar to this:

```
fun AsyncInputStream.consume(autoclose: Boolean = true, block: suspend (data: ByteArray, offset: Int, size: Int) -> Unit) {
    try {
    val temp = ByteArray(1024)
    while (true) {
          val read = read(temp, 0, temp.size)
          if (read < 0) break
          block(temp, 0, read)
    }
    } finally {
          if (autoclose) close()
    }
}
```

And write around that something for pipeing without pressure, like:

```
fun AsyncInputStream.pipeTo(out: AsyncOutputStream) {
     consume { data, offset, size -> out.write(data, offset, size) }
}
```

Is that the kind of higher level functions you were talking about? If so, I think we can add them to the library if they are not there already
